extensions/transaction/transaction-atomikos/src/main/java/org/eclipse/dataspaceconnector/transaction/atomikos/AtomikosTransactionContext.java

```diff
@@ -66,17 +66,49 @@ public void execute(TransactionBlock block) {
             }
             throw new EdcException(e);
         } finally {
-            if (startedTransaction) {
-                try {
-                    var status = transactionManager.getStatus();
-                    if (STATUS_ACTIVE == status) {
-                        transactionManager.commit();
-                    } else if (STATUS_MARKED_ROLLBACK == status) {
-                        transactionManager.rollback();
-                    }
-                } catch (HeuristicRollbackException | SystemException | HeuristicMixedException | RollbackException e) {
-                    monitor.severe("Transaction error", e);
+            commitOrRollback(startedTransaction);
+        }
+    }
+
+    @Override
+    public <T> T execute(ResultTransactionBlock<T> block) {
+        var startedTransaction = false;
+        Transaction transaction = null;
+        try {
+            transaction = transactionManager.getTransaction();
+            if (transaction == null) {
+                transactionManager.begin();
+                transaction = transactionManager.getTransaction();
+                startedTransaction = true;
+            }
+
+            return block.execute();
+
+        } catch (Exception e) {
+            try {
+                if (transaction != null) {
+                    transaction.setRollbackOnly();
+                }
+            } catch (SystemException ex) {
+                monitor.severe("Error setting rollback", ex);
+            }
+            throw new EdcException(e);
+        } finally {
+            commitOrRollback(startedTransaction);
+        }
+    }
+
+    private void commitOrRollback(boolean startedTransaction) {
+        if (startedTransaction) {
+            try {
+                var status = transactionManager.getStatus();
+                if (STATUS_ACTIVE == status) {
```

Comment:
Implementing a case for other statuses would probably be good.

Reply:
Can't really comment, that would probably be a topic for @jimmarino. In any event I think it's out of scope for this PR. It only turns up as changed because i added the execute-with-result temporarily.